src/dodal/devices/detector/detector.py

```diff
@@ -41,44 +46,40 @@ class DetectorParams(BaseModel):
     det_dist_to_beam_converter_path: str
     trigger_mode: TriggerMode = TriggerMode.SET_FRAMES
     detector_size_constants: DetectorSizeConstants = EIGER2_X_16M_SIZE
-    beam_xy_converter: DetectorDistanceToBeamXYConverter = None  # type: ignore # Filled in by validator
-    run_number: int = None  # type: ignore # Filled in by validator
     enable_dev_shm: bool = (
         False  # Remove in https://github.com/DiamondLightSource/hyperion/issues/1395
     )
+    model_config = ConfigDict(
+        arbitrary_types_allowed=True,
+    )
 
-    class Config:
-        arbitrary_types_allowed = True
-        json_encoders = {
-            DetectorDistanceToBeamXYConverter: lambda _: None,
-            DetectorSizeConstants: lambda d: d.det_type_string,
-        }
-
-    # should be replaced with model_validator once move to pydantic 2 is complete
-    @root_validator(pre=True)
-    def create_beamxy_and_runnumber(cls, values: dict[str, Any]) -> dict[str, Any]:
-        values["beam_xy_converter"] = DetectorDistanceToBeamXYConverter(
-            values["det_dist_to_beam_converter_path"]
-        )
-        if values.get("run_number") is None:
-            values["run_number"] = get_run_number(values["directory"], values["prefix"])
-        return values
-
-    @validator("detector_size_constants", pre=True)
-    def _parse_detector_size_constants(
-        cls, det_type: str, values: dict[str, Any]
-    ) -> DetectorSizeConstants:
+    @property
+    def beam_xy_converter(self) -> DetectorDistanceToBeamXYConverter:
+        return DetectorDistanceToBeamXYConverter(self.det_dist_to_beam_converter_path)
+
+    @property
+    def run_number(self) -> int:
+        return get_run_number(self.directory, self.prefix)
```

Comment:
Must: This now means that `run_number` cannot be specified when creating a `DetectorParams`. This is why some unit tests are failing, including `test_when_set_odin_pvs_called_then_full_filename_written_and_set_mx_settings_runs` as it expects odin to be set to create a file with the run number it has specified rather that trying to find one from disk. I think we still want to keep the behaviour of being able to specify the `run_number`

Reply:
Does being able to specify `override_run_number: int | None = None`, which is then returned if set by the property `run_number() -> int` else the get_run_number function suffice? I think explicitly naming it override makes it clearer what the intended behaviour is, but still leaves it as optional and has no # type: ignore to worry about.

Reply:
Yes, but how would you feel about spinning this into a new issue as it would involve changes in hyperion and GDA to get to work?

Reply:
If setting a JSON alias so that the blob from GDA is deserialised from `"run_number": <override_run_number>` correctly, would that suffice? This handling is much simpler than dealing with field validators to optionally inject the value: I don't want to have to roll this back

Reply:
I've added a new test to ensure that the behaviour is as expected: DetectorParams.override_run_number serialises and deserialises as "run_number".

Reply:
https://github.com/DiamondLightSource/dodal/issues/775

Reply:
Great, thank you! Think I'm happy to approve now, once tests/linting is happy. I will try and do the `mx-bluesky` side today